docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,384 @@
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+**Table of Contents**
+
+- [Overview](#overview)
+- [Cascading deletion with Garbage Collector](#cascading-deletion-with-garbage-collector)
+- [Orphaning the descendants with "orphan" finalizer](#orphaning-the-descendants-with-orphan-finalizer)
+  - [Part I. The finalizer framework](#part-i-the-finalizer-framework)
+  - [Part II. The "orphan" finalizer](#part-ii-the-orphan-finalizer)
+- [Related issues](#related-issues)
+  - [Orphan adoption](#orphan-adoption)
+  - [Upgrading a cluster to support cascading deletion](#upgrading-a-cluster-to-support-cascading-deletion)
+- [End-to-End Examples](#end-to-end-examples)
+  - [Life of a Deployment and its descendants](#life-of-a-deployment-and-its-descendants)
+  - [Life of a ConfigMap](#life-of-a-configmap)
+- [Open Questions](#open-questions)
+- [Considered and Rejected Designs](#considered-and-rejected-designs)
+- [1. Tombstone + GC](#1-tombstone--gc)
+- [2. Recovering from abnormal cascading deletion](#2-recovering-from-abnormal-cascading-deletion)
+
+
+# Overview
+
+Currently most cascading deletion logic is implemented at client-side. For example, when deleting a replica set, kubectl uses a reaper to delete the created pods and then delete the replica set. We plan to move the cascading deletion to the server to simplify the client-side logic. In this proposal, we present the garbage collector which implements cascading deletion for all API resources in a generic way; we also present the finalizer framework, particularly the "orphan" finalizer, to enable flexible alternation between cascading deletion and orphaning.
+
+Goals of the design include:
+* Supporting cascading deletion at the server-side.
+* Centralizing the cascading deletion logic, rather than spreading in controllers.
+* Allowing optionally orphan the dependent objects.
+
+Non-goals include:
+* Releasing the name of an object immediately, so it can be reused ASAP.
+
+# Cascading deletion with Garbage Collector
+
+## API Changes
+
+```
+type ObjectMeta struct {
+	...	
+	OwnerReferences []OwnerReference
+}
+```
+
+**ObjectMeta.OwnerReferences**:
+List of objects depended by this object. If ***all*** objects in the list have been deleted, this object will be garbage collected. For example, a replica set `R` created by a deployment `D` should have an entry in ObjectMeta.OwnerReferences pointing to `D`, set by the deployment controller when `R` is created. This field can be updated by any client that has the privilege to update the object. For safety reasons, we can add validation rules to restrict what resources could be set as owners. For example, Events will likely be banned from being owners.
+
+```
+type OwnerReference struct {
+    // Group of the referent.
+    Group string
```

Comment:
Why is there a group here? Are we going to introduce group everywhere we have apiVersion now?